### ISSUE TITLE
Document comma-dangle: functions: always-multiline behaviour in promises

### ISSUE THREAD

**Issue by mcky** [eslint/eslint#7951]
1. **ESLint Version**
v3.13.1 (although this is documentation specific, not the linter itself

2. **The problem you want to solve.**
The behaviour of always-multiline comma dangle in promises isn't clearly documented. As In #7786 @mysticatea [described why](https://github.com/eslint/eslint/issues/7786#issuecomment-267898825) a trailing comma is forced in some `Promise.then` situations

3. **Your take on the correct solution to problem.**
Update the documentation for the rule to include a `always-multiline` section with the rest of the function docs, showing:
```diff
-Promise.then(x => x
-  .replace('asdasdsdsda', '123213213')
-  .replace('cxvcxvcxvcx', '546546546')
-  .replace('ewrewrewrew', '877876876'),
-)
+Promise.then(x => (
+  x.replace('asdasdsdsda', '123213213')
+    .replace('cxvcxvcxvcx', '546546546')
+    .replace('ewrewrewrew', '877876876')
+))
```

Raised this as an issue instead of a PR as I guess it's a conflict between two rules in the airbnb config, but it might not belong on that page

**Comment by not-an-aardvark** [comment]
Thanks for the report. Could you clarify what part of the behavior was confusing? The rule doesn't have any special behavior for Promises -- this is just how it behaves with multiline arguments. I'm trying to understand what the confusion is so that the docs can be clarified most effectively.

**Comment by mcky** [comment]
Ah I hadn't thought of that.. Yeah sorry I should've specified with chaining not promises, although it's a common pattern with promises compared to regular callbacks.

I guess the confusion comes from the rule causing syntax errors, but only being fixable by re-arranging or wrapping in `()`

**Comment by platinumazure** [comment]
@not-an-aardvark Are trailing commas a syntax error if the last argument is an arrow function (possibly due to ambiguity over whether arrow function is returning a sequence in that case)?

**Comment by not-an-aardvark** [comment]
@platinumazure I'm not sure, but I don't think so -- arrow function expressions have a higher precedence than sequence expressions.

**Comment by mcky** [comment]
```
Promise.resolve([1, 2, 3])
.then(x => x
	.map(y => y),
)
```
<img width="542" alt="screen shot 2017-01-20 at 17 15 21" src="https://cloud.githubusercontent.com/assets/2717635/22158434/14f5afae-df34-11e6-8ee2-5e525c35ac7f.png">
<img width="322" alt="screen shot 2017-01-20 at 17 18 55" src="https://cloud.githubusercontent.com/assets/2717635/22158532/93308074-df34-11e6-9d22-da7b836f7f3f.png">

**Comment by not-an-aardvark** [comment]
@mcky Trailing function commas are an ES2017 feature which has not been implemented in Chrome yet, so Chrome will throw a syntax error for all trailing function commas regardless of whether they contain arrow functions. I think @platinumazure was asking whether arrow functions would cause a syntax error according to the ES2017 spec (with the assumption that the given implementation supports trailijng commas in the general case).

**Comment by mcky** [comment]
Ok, I'll definitely update our config but is there a way to make it clearer to others trying to figure out the same problem?
I mentioned it to someone on our team and they said they'd ran into the same thing and didn't find a solution.

Would a notice about it being es2017 around the functions section make it clearer? 

Stylistically I can't say I see the benefit of comma dangle in functions, needing them for diffs would be a sign of needing to refactor to destructuring or something - but that's not really pertinent to this discussion

**Comment by not-an-aardvark** [comment]
> Would a notice about it being es2017 around the functions section make it clearer?

I agree, this seems like it would help a lot. Would you like to make a PR?

Also, keep in mind that the parser configuration (e.g. `ecmaVersion: 2016` and the rule configuration (e.g.  `functions: always-multiline`) are independent from each other. The default `functions` configuration is `ignore` (i.e. no errors are reported for trailing function commas), but if you're using a shareable config such as `eslint-config-airbnb`, it's possible the config is explicitly enabling `functions: always-multiline`. To override this, you can use something like

```js
rules: {
  'comma-dangle': ['error', 'always-multiline']
}
```

**Comment by merlinpatt** [comment]
This same issue happens when mapping as well.

```
{units.map(unit =>
  <UnitTile key={unit._id} course={course} unit={unit} />
)}
```

This rule recommends I add a comma after the `/>`, which then becomes a parse error on the next run through.

There does not seem to be any benefit that I can think of that comes from adding a comma here because there is nothing else that would go after the returned component.

**Comment by not-an-aardvark** [comment]
This is fixed by https://github.com/eslint/eslint/pull/8468.
